Enable AWS Parallel Works platform and Add Comprehensive End-To-End Tests

jessemcfarland opened the pull request:
## DESCRIPTION OF CHANGES: 
First, the AWS Parallel Works platform has been activated. A couple additional minor changes were necessary in order to get a successful build and test run. These changes include mapping the Parallel Works cluster names to `noaacloud` and ensuring the the `PROJ_LIB` environment variable is set. 

Second, support for the comprehensive workflow/end-to-end tests was added to the Jenkins pipeline and unified test script. A boolean parameter, `SRW_WE2E_COMPREHENSIVE_TESTS`, was added to the Jenkins pipeline. The parameter can be used to execute the comprehensive test suite manually on the desired branches. In addition, logic was added to the test stage to scan Pull Request labels for a specific label, `run_we2e_comprehensive_tests`. If set, the value of the  `SRW_WE2E_COMPREHENSIVE_TESTS` parameter is overridden. The list of comprehensive workflow/end-to-end tests was added to the unified test script. 

## TESTS CONDUCTED: 
This PR is the test. Two new labels should be created: `run_we2e_default_tests` and `run_we2e_comprehensive_tests`. First, the`run_we2e_default_tests` label should be applied to the PR. This should result in the pipeline executing the default set of tests. Next, the `run_we2e_default_tests` label should be removed and the `run_we2e_comprehensive_tests` label should be added. This should result in the pipeline executing the comprehensive set of tests. __NOTE__: The tests results can be found under the __S3 Artifacts__ section of a Jenkins __Build__.

Commit:
Normalize Parallel Works cluster platform value

Set the value of platform to 'noaacloud' when SRW_PLATFORM matches a
Parallel Works cluster name.

Commit:
Enable the AWS Parallel Works platform

Commit:
Move agent declaration to stages

This change allows the platform filter to work correctly, otherwise, the
Parallel Works clusters would block indefinitely waiting to execute the
matrix on a agent/node that was not started.

Commit:
Ensure PROJ_LIB is set on Parallel Works platforms

Commit:
Fix final exit status of srw_test script

Some platforms do not recognize quoted variables within an arithmetic
expression. This change removes the quotes.

Commit:
Add comprehensive end-to-end tests option

* Add a parameter to the Jenkins pipeline that allows the comprehensive
set of workflow and end-to-end tests to be executed during the test
stage.
* Add logic to the Jenkins pipeline that checks for a specific Pull
Request label, then overrides the comprehensive end-to-end test
parameter's value if set.

Commit:
Clean up the workspace after a we2e test run

The experiments directory uses a lot of disk space. Removing it after
the end-to-end tests complete will allow us to keep the workspaces
longer. However, the test logs should be preserved. This change creates
a tarball containing the test logs in the workspace, which is archived,
then removes the experiments directory.

JeffBeck-NOAA commented:
@jessemcfarland, do you mind changing "default" to "fundamental"?  We're using the terms "comprehensive" to describe the full suite of WE2E tests, and "fundamental" to describe the WE2E tests that should be run for each PR.  Thanks!

jessemcfarland commented:
> @jessemcfarland, do you mind changing "default" to "fundamental"? We're using the terms "comprehensive" to describe the full suite of WE2E tests, and "fundamental" to describe the WE2E tests that should be run for each PR. Thanks!

@JeffBeck-NOAA Done!

JeffBeck-NOAA reviewed (approved):
Thanks, @jessemcfarland!

jessemcfarland commented:
I aborted the tests early, but the labels are working as expected.

gsketefian commented on .cicd/scripts/srw_test.sh:
```diff
@@ -17,19 +17,72 @@ else
-we2e_default_tests=('grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16'
+# The fundamental set of end-to-end tests to run.
+declare -a we2e_fundamental_tests
+we2e_fundamental_tests=('grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16'
```
The fundamental tests should also include these two:
```
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_HRRR
```
See [here](https://docs.google.com/spreadsheets/d/18CO_OtLsLeRBMcW0O8YdS4ipancKva0Qospz1xaWIdY/edit#gid=0).  Tests for the `WoFS` suite should also be added as soon as the problems with that suite are worked out.

EdwardSnyder-NOAA replied:
The `WoFS` suite has been added, so you can add the below test to the fundamental test list as well: `grid_SUBCONUS_Ind_3km_ics__HRRR_lbcs_RAP_suite_WoFS_v0`.

Also, the HRRR suite tests are labeled as 'suite_HRRR' not 'suite_RRFS_HRRR'. So this test `grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_HRRR` needs to be `grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_HRRR`.

gsketefian replied:
@EdwardSnyder-NOAA Thanks for the correction on `grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_HRRR`.

EdwardSnyder-NOAA replied:
Just noticed there is a typo in the WoFS fundamental test. There is an extra '_' between 'ics' and 'HRRR'. So the WoFS test should be: `grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_WoFS_v0`.

gsketefian replied:
@EdwardSnyder-NOAA I don't see a WoFS test in this script (`srw_test.sh`) yet.

jessemcfarland replied:
@gsketefian Just pushed the update now, testing now.

jessemcfarland replied:
The three new tests failed, but I think that's expected right now? They ran, so the script update is working correctly at least. At this point I think this PR is good to go.

gsketefian replied:
@jessemcfarland It looks like the `WoFS_v0` test failed because it couldn't find the external model files.  That usually means either the file naming convention wasn't properly specified or the file isn't there on disk.  As you say, it's not related to this PR.

The `RRFS_v1beta` and `HRRR` tests failed in the `make_lbcs` task because a (temporary?) file couldn't be found.  Also not directly related to this PR (it probably has to do with ESMF versions), but it did remind me that there are other log files that should be included in the Jenkins artifacts.  These are the files in, e.g., `$EXPTDIR/2019070100/INPUT/tmp_ICS` and `$EXPTDIR/2019070100/INPUT/tmp_LBCS`.  The most important of these are named `PET##.ESMF_LogFile` (they are ESMF log files for each MPI task it runs, where ## is the task number.  These are usually needed when the `make_ics` or `make_lbcs` tasks fail.

If it's not too much trouble, it would be great to include those PET files as part of this PR; otherwise, in another PR is fine.  Thanks.

EdwardSnyder-NOAA replied:
@gsketefian - there was a bug in the code that caused the `RRFS_v1beta` and `HRRR` tests to fail in the `make_lbcs` step. This has been addressed in [PR 795](https://github.com/ufs-community/regional_workflow/pull/795), which was merged in earlier today. As for the `WoFS_v0` test failing, this is something we are aware of and will be testing the solution (adding WoFS to `ufs-srweather-app/src/CMakeLists.txt`) soon.

gsketefian commented on .cicd/scripts/srw_test.sh:
```diff
@@ -17,19 +17,72 @@ else
+
+# The set of workflow tests to run.
+declare -a we2e_workflow_tests
+we2e_workflow_tests=('community_ensemble_008mems'
```
Specifying the lists of tests explicitly in this script is ok for the `release/public-v2` branch, but when this is generalized to the `develop` branch, the tests to be run should be read in from some external source (e.g. a file).  That's because a PR itself may change the tests (change test names and add/delete tests).  That more dynamic approach requires adding a test type to each test, e.g. by creating a variable in each test config file named `WE2E_TEST_TYPE` that can be set to `fundamental` or `comprehensive`.  So this script can call another script (e.g. some modified version of `get_WE2Etest_names_subdirs_descs.sh` in `regional_workflow/tests/WE2E`) that outputs a text file identifying the type of each test (fundamental or comprehensive).  This script can then read in that file to determine which tests to run.  

This is just a note for when the automated tests are added to the `develop` branch.  Doesn't affect this PR.

gsketefian replied:
I should also add that when the tests are added to the `develop` branch, there needs to be documentation added that describes how to run the tests and a bit of what's happening under the hood (which scripts do what).

gsketefian replied:
One more thing, again for when this is generalized to `develop`.  In the PR message, it says "NOTE: The tests results can be found under the S3 Artifacts section of a Jenkins Build."  Does that mean the user needs a Jenkins account that's somehow linked to this repo?  If so, I imagine that can be a bit of a pain to access.  What report comes back to github?  It would be great if at a minimum the list of which tests passed and failed were returned, even better if the the `log.launch_FV3LAM_wflow` log file for each test was returned, and best if the log files from each task were returned so the user can see what exactly went wrong during the test (those log files are under the `${test_name}/log` directory of each test).

jessemcfarland replied:
@gsketefian the users don't need a Jenkins account to view the build log and artifacts. I am capturing the `log.launch_FV3LAM_wflow` in this PR and am working on an update to capture the `log.generate_FV3LAM_wflow` now. I can also include the files under the `${test_name}/log` directory. I may need to put them in all in a tarball for each test to keep things relatively clean. How does that sound?

gsketefian replied:
@jessemcfarland That sounds excellent :). I wonder if it would be better to have all the test log files in one tarball instead of each test in a separate one.  That way the user needs to `untar` only one file.  Just a thought, I don't know the answer.  Also, I'm thinking It would be good to have the file that summarizes the final status of each test outside the tarball(s) so the user can just look at it without the need to `untar`.  But we can first try whatever you think is best and make adjustments later as needed.  Thanks for your work on this.

jessemcfarland replied:
I just pushed this update and am testing it now. I did one tar file per platform/compiler combo. It'll be in the S3 Artifacts as `we2e_test_logs-<PLATFORM>-<COMPILER>.tgz`. We do have a test results file that just has the test name and final status, which is also in the S3 Artifacts as `we2e_test_results-<PLATFORM>-<COMPILER>.txt`.

jessemcfarland replied:
@gsketefian you should be able see the artifacts for this latest build here: https://jenkins-epic.woc.noaa.gov/job/ufs-srweather-app/job/pipeline/job/PR-297/9/s3/. I ran only on Jet for simplicity, but for a normal run you'd see the same three files for each platform/compiler combo.

gsketefian replied:
@jessemcfarland Thanks.  Couple of questions:
1) Will the url to click on to get to those test results be returned somewhere on the PR page for users to click on?  I checked towards the bottom of this page but didn't see the url you sent me anywhere else.
2) How long are these artifacts kept on jenkins before being scrubbed?
3) `ufs-srweather-app_jet-intel-PR-297-9.tgz` seems to contain the files that the build generates, is that right?  If so, why not also include the `build` directory in it?

jessemcfarland replied:
@gsketefian 
1. You can reach the pipeline display page by clicking **Details** next to **Build and Test** under the checks section of the PR status below. From that page, you need to switch from the _Blue Ocean_ UI to _Classic_ by clicking the _->_ icon in the top right corner of the page. From the _Classic_ build page, you'll find **S3 Artifacts** on the left navigation menu, assuming there are artifacts for the build.
2. This is configurable, so we can do whatever makes sense. I think this is one topic to discuss in a larger discussion we should have on how the pipeline can be used for the development and release workflows. Right now, the artifacts for the `develop` and release branches are kept indefinitely and artifacts for defunct branches and PRs are kept for 30 days.
3. The intent was to preserve the binaries created by a build so that can be used on the target platforms. My understanding is that the `build` directory is just the "working directory" for the build and wasn't required to use the binaries, is that not the case? You can still examine the `build` directory for the latest build by inspecting the Jenkins _Workspace_.

gsketefian replied:
@jessemcfarland:
1. Thanks, I found them.  It would be great if those steps are documented somewhere (tagging @gspetro-NOAA).  Do artifacts from all PRs go into the same S3 bucket? I.e. will they be in one big list or does each PR have its own bucket/subdirectory/storage area?
2. Ok, we can discuss it during the next meeting (which will likely not be this week).  I put it [here](https://docs.google.com/document/d/1HS7u5xo9VoS0Xh2BR4gMUXKgtYkwXJtHWllYotEeiFQ/edit) in the meeting notes.
3. You're right, the build directory is not needed for using the binaries.  I was thinking it may have log files from the build, but actually I don't think it does.  Is the output of the build saved anywhere to see what went wrong if the build fails?  I usually just redirect it to a file.
4. I found the Jenkins Workspace using the link above for the artifacts.  Looks like it's a copy of the ufs-srweather-app directory, is that right?  I found 3 workspaces that are called this:
[/scratch1/NCEPDEV/stmp2/role.epic/jenkins/workspace/fs-srweather-app_pipeline_PR-297](https://jenkins-epic.woc.noaa.gov/job/ufs-srweather-app/job/pipeline/job/PR-297/12/execution/node/113/ws/) on [Hera](https://jenkins-epic.woc.noaa.gov/computer/Hera)
  Since there is no date or other distinct number in this (except the PR #, which is the same for all three), it's hard to say which represents the latest set of tests.  Any suggestions?

Thanks again for all your work on this.  It's a much needed capability.

jessemcfarland replied:
@gsketefian:
1. Agreed on the docs! Yes, all artifacts go to the same S3 bucket right now, but it is possible to use more than one if we decide that's beneficial. Artifacts are organized by job and build. For multi-branch pipeline jobs, like this one, they'll be organized by job, branch/change request, and build. 
2. Great! I may need an invite for that one.
3. OK thanks! The output of the build is saved by Jenkins. The best way to get there is from the _Blue Ocean_ UI (the page that comes up when you click the build details below): each row in the pipeline is a platform/compiler combo (you can mouse over _Matrix_ to see the params): click on the **Build** node for the platform/compiler combo you want to view, that'll bring up the steps for that stage at the bottom of the page, then click on the step that runs the `srw_build.sh` script, which will bring up the output of the build. We can also have the build script dump the build output to a file and archive it along with the other artifacts.
4. That's right! The workspace is Jenkins' working directory, so it's where it clones the repo, performs the build/test, etc. The files can be examined during or after a build. The _Workspace_ link in the UI is a pointer to the actual workspace, meaning it's how Jenkins sees it at this moment and is always pointed at the latest build. So anything that should be preserved between builds needs to be archived.

gsketefian replied:
@jessemcfarland:
1. Thanks, I was able to look at the 7 pipeline branches (not to be confused with git branches!) and look at the details under each.  For the documentation, I created an issue in this repo (#[307](https://github.com/ufs-community/ufs-srweather-app/issues/307)) and took the liberty to assgin you and @gspetro to it.  Please let me know (either of you) if that's not ok.
2. I invited you to the Thursday code management meetings (indefinitely for now; can remove you later if you'd like).  No meeting this Thursday, but there should be one next week.
3. Found the build log under Matrix.  Yeah, it would be good to archive it as well so a user can download everything and view it locally instead of having to toggle back and forth between their machine and Jenkins.  Feel free to do it in this PR or another one.
4. Ok.  Don't quite understand why there are three lines named the same (see attachment).  I have to play with this more to understand it better.
<img width="972" alt="Jenkins_workspaces" src="https://user-images.githubusercontent.com/31046882/173874348-c337a810-0bb5-4b84-9690-be575b2e71fb.png">

jessemcfarland replied:
@gsketefian ah ok, that's a good question, I believe there is a workspace link for each stage in the pipeline, init, build, test. For this pipeline, they all happen to point to the same workspace and machine. This would be a more handy feature in another use case where you build on one machine and test on a different machine.

Commit:
Disable concurrent builds for branches and PRs

Prevent Jenkins from executing multiple pipelines at the same time for a
given branch or change request.

Commit:
Disable branch indexing triggers for pipeline

Commit:
Update we2e fundamental tests for srw_test script

gsketefian commented:
@jessemcfarland To summarize my comments, it would be nice to have the following two items archived as well, but feel free to do it in another PR if you like.
1. The log of the build.
2. The `PET##.ESMF_LogFile` files that ESMF generates while running the `make_ics` and `make_lbcs` tasks.

Commit:
Log and archive the output of the srw build

jessemcfarland commented:
REBUILD

jessemcfarland commented:
REBUILD

Commit:
Update we2e comprehensive tests in srw_test script
